### PR TITLE
feat(eagle3): on-policy regeneration loop with step-cadence dataloader swap

### PR DESCRIPTION
**Pull request by khazic**
## Summary

Adds the on-policy data regeneration loop for EAGLE-3 draft training ("train-with-decode"), on top of the periodic decode-eval that landed in #3037.

EAGLE-3 teacher-forces the draft on a static dataset whose assistant turns were often written by a different model than the target being distilled, so the draft trains against off-distribution supervision. This feature regenerates that data online, interleaved with training:

- On a step cadence (`regen.every_steps`), a detached rank-0 worker subprocess boots a plain vLLM server for the frozen target on a reserved GPU and regenerates a fresh directory of shards. Training never blocks; at most one cycle is in flight.
- At a grad-accum window boundary (gradients clean, no partial window to flush), all data-parallel ranks rebuild the train dataloader against the new shard directory in lockstep, with the decision broadcast from rank 0. The epoch loop is now step-budget driven, so a single-epoch run trains on regenerated data instead of no-opping.

## Correctness properties

Surfaced by an adversarial code-review pass and fixed:

- Per-segment `batches_per_epoch` recompute after a swap, so `should_sync_grads` and the trailing-window flush stay correct. A stale count would run the flush on un-all-reduced gradients and silently desync the DDP replicas.
- `take_ready_shards` frees every superseded ready cycle, not only the previously-active one, so disk does not grow one dataset per cycle on the shared filesystem.
- On resume, the launch cadence is aligned to the restored `global_step` (`resume_from_step`), so a resumed run does not relaunch a redundant cycle for an already-covered region.
- A crashed cycle's directory is wiped before relaunch, since regenerate aborts on a non-empty shard dir.
- The LR horizon is kept fixed (the step-budget stop keeps the cosine coherent) and warns if a regenerated cycle's length differs from the original.
- A `regen.output_dir` override is rejected loudly instead of being silently discarded.

## Tests

- `tests/unit_tests/speculative/test_regen_loop.py` (28): config resolution and validation, cadence / consume / GC, shutdown signal escalation, resume alignment, worker argv and `main`. 99% line coverage of `regen_loop.py` (only the `__main__` guard uncovered).
- `tests/unit_tests/recipes/llm/test_train_eagle3_regen.py` (13): `_setup_regen` gating, `_build_train_dataloader` forwarding, the swap path (rank-0 decide, rebind, LR-drift warning), the step-budget segment loop swapping mid-epoch, prefetch drain on early break, and regen-off backward compatibility.
- The existing eagle recipe suite (159 tests) still passes.

## GPU validation status

A 2-rank colocated smoke run caught and fixed a real setup-ordering bug: `_setup_regen` read `self.output_dir` before the setup body assigned it, and a unit regression test now covers it. The regen code path is validated through setup, where distributed init, target load, and runner construction all succeed.

Full end-to-end training plus swap plus DDP gradient-sync validation is still pending. The base EAGLE-3 answer-only loss-mask preprocessing currently fails for Qwen3 (it asks for a chat template that wraps assistant turns in generation tags), which reproduces with no regen block and is therefore orthogonal to this feature. Keeping the PR as a draft until that is resolved and the full run is validated.

### CODE REVIEW

File: nemo_automodel/recipes/llm/train_eagle3.py [hunk 1]
```diff
@@ -1639,102 +1761,137 @@ def _train_epochs(self, start_epoch, batches_per_epoch, is_ddp, pbar=None):
...
-                        running_loss.zero_()
-                        running_acc.zero_()
-                        if running_step_prefix is not None:
-                            running_step_prefix.zero_()
-                            running_step_valid.zero_()
-                        running_steps = 0
+                                    wandb_data["train/tau_sim"] = tau_sim
+                                self._wandb_log(wandb_data, step=self.runtime.global_step)
+                                self._maybe_run_decode_eval()
+                                self._maybe_run_regen()
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/recipes/llm/train_eagle3.py]: Good catch, fixed in 69d0153: the cadence check now runs at every grad-accum window boundary (after the budget break, before the swap check), with a unit test pinning launch checks at steps 1..3 under log_every_steps=100.

**Comment by khazic** [comment]
## End-to-end validation (single-GPU smoke)

Ran the on-policy regen loop end-to-end on one A100-80GB, target `Qwen/Qwen3-4B` (colocated). Only one card was free, so the regen vLLM server was co-located on the same card as training (`regen.cuda_visible_devices` pointed at the training card) — enough to exercise the full path: cadence launch → the worker boots a plain vLLM OpenAI server for the frozen target → `regenerate` writes a fresh shard dir + `READY` → the recipe swaps the train dataloader at the next grad-accum window boundary.

Regen block used:

```yaml
regen:
  every_steps: 3
  cuda_visible_devices: "0"          # co-located with training for the smoke
  server_python: <vllm 0.23.0 env python>
  input_data: <single-turn slice of PerfectBlend messages>
  split: "train[:8]"                 # tiny on purpose
  shard_size: 8
  max_new_tokens: 32
  gpu_memory_utilization: 0.30
  max_model_len: 4096
```

What the run confirms:

- **cadence launch** — `regen: launched cycle for step 3 on CUDA_VISIBLE_DEVICES=0`, i.e. the worker fired at the configured `every_steps` (and, per the fix in 69d0153, the launch check now runs at every optimizer-step boundary, so it is independent of `log_every_steps`).
- **one-in-flight invariant** — every due-check while the worker was still running logged `previous cycle (step 3) still running, skipping launch at step N`.
- **worker booted a vLLM server for the frozen target** and drove `regenerate` to completion (`regen worker: cycle 3 ready at .../shards`; the shard dir holds `shard-000000.parquet` + `manifest.json` + the `READY` marker).
- **dataloader swap** — `regen: swapping train dataloader to regenerated shards` fired at the next window boundary, and the length guard correctly warned `regenerated dataloader has 8 batches vs the original 300` (expected here: the smoke regenerates only 8 prompts).
- **training continued post-swap with no errors**, and the vLLM server shut down cleanly when the cycle ended.

(Data note: the smoke trains on a single-turn slice of PerfectBlend with `Qwen/Qwen3-4B`. Qwen3's chat template rewrites earlier assistant turns, so the answer-only fallback mask now fails loud on multi-turn rows — the fix from #3024 — which is orthogonal to this PR; a single-turn slice sidesteps it.)

<details><summary>Sanitized log — key phases</summary>

```text
# --- setup / draft vocab scan ---
INFO | nemo_automodel.components.datasets.llm.eagle3 | Counted 44075 supervised tokens across 3614 unique target tokens
INFO | nemo_automodel.recipes.llm.train_eagle3 | Training start: start_epoch=0 num_epochs=40 batches_per_epoch=300 grad_accum=1 log_every=5 total_optim_steps=12000 warmup_steps=600 peak_lr=2.000e-04 min_lr_ratio=0.1

# --- regen worker launched at the cadence (step 3) ---
INFO | nemo_automodel.components.speculative.regen_loop | regen: launched cycle for step 3 on CUDA_VISIBLE_DEVICES=0 (pid ..., log /data/runs/eagle3_regen_e2e/out/regen/cycle_3/worker.log)
INFO | nemo_automodel.components.speculative.regen_loop | regen: previous cycle (step 3) still running, skipping launch at step 6
INFO | nemo_automodel.components.speculative.regen_loop | regen: previous cycle (step 3) still running, skipping launch at step 7
... (repeats every step while the worker runs) ...

# --- worker: boot vLLM OpenAI server for the frozen target ---
regen worker: starting target server: python -m vllm.entrypoints.openai.api_server --model Qwen/Qwen3-4B --served-model-name target --host 127.0.0.1 --port ##### --gpu-memory-utilization 0.3 --max-model-len 4096
(EngineCore) Initializing a V1 LLM engine (v0.23.0) ... model='Qwen/Qwen3-4B', enforce_eager=False, ...
(EngineCore) Loading weights took 3.20 seconds
(EngineCore) Model loading took 7.56 GiB and 5.40 seconds
(EngineCore) torch.compile took 5.83 s in total          # warm compile cache
(EngineCore) Available KV cache memory: 15.84 GiB
(EngineCore) GPU KV cache size: 115,344 tokens
(EngineCore) Graph capturing finished in 8 secs, took 0.58 GiB
regen worker: cycle 3 ready at /data/runs/eagle3_regen_e2e/out/regen/cycle_3/shards

# --- recipe: swap the train dataloader to the regenerated shards ---
INFO | nemo_automodel.recipes.llm.train_eagle3 | regen: swapping train dataloader to regenerated shards at /data/runs/eagle3_regen_e2e/out/regen/cycle_3/shards
WARNING | nemo_automodel.recipes.llm.train_eagle3 | regen: regenerated dataloader has 8 batches vs the original 300; the fixed LR horizon (total_optim_steps) assumes an unchanged per-pass step count, so the cosine schedule may drift.

# --- training continues on the regenerated data (no errors) ---
epoch=37 step=1795 train_loss=0.340622 train_acc=0.962891 train_tau_sim=4.5640 lr=1.952e-04
epoch=38 step=1800 train_loss=0.083337 train_acc=1.000000 train_tau_sim=5.0000 lr=1.951e-04
epoch=39 step=1810 train_loss=0.238417 train_acc=0.977865 train_tau_sim=4.7086 lr=1.950e-04
(high acc/tau here just reflect overfitting the 8-sample regenerated slice — this is a plumbing smoke, not a convergence run)

# --- vLLM server shut down cleanly at the swap ---
(APIServer) [shutdown] API server: shutdown triggered
(APIServer) [shutdown] API server: shutting down FastAPI HTTP server
(APIServer) Application shutdown complete.
```
</details>

**Review by HuiyingLi — APPROVED** [review]
